WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -299,28 +306,102 @@ async Task ReadyToSignTask(AliceClient aliceClient)
-		// * bigger coins first (this makes economical sense because mix more money paying less network fees)
-		//
-		// Note: this method works on already pre-filteres coins: those available and that didn't reached the
-		// expected anonymity set threshold.
-		private ImmutableList<SmartCoin> SelectCoinsForRound(IEnumerable<SmartCoin> coins, MultipartyTransactionParameters parameters) =>
-			coins
+		private ImmutableList<SmartCoin> SelectCoinsForRound(IEnumerable<SmartCoin> coins, MultipartyTransactionParameters parameters)
+		{
+			var filteredCoins = coins
 				.Where(x => parameters.AllowedInputAmounts.Contains(x.Amount))
 				.Where(x => parameters.AllowedInputTypes.Any(t => x.ScriptPubKey.IsScriptType(t)))
-				.GroupBy(x => x.TransactionId)
-				.Select(x => x.OrderByDescending(y => y.Amount).First())
+				.ToShuffled() // Preshuffle before ordering.
 				.OrderBy(x => x.HdPubKey.AnonymitySet)
-				.ThenByDescending(x => x.Amount)
-				.Take(MaxInputsRegistrableByWallet)
-				.ToShuffled()
-				.ToImmutableList();
+				.ThenByDescending(y => y.Amount)
+				.ToArray();
+
+			// If there's no non-private coins then there's no reason to mix, the rest that's here has already reached the min anonscore target.
+			int nonPrivateCoinCount = filteredCoins.Where(x => x.HdPubKey.AnonymitySet < MinAnonScoreTarget).Count();
+			if (nonPrivateCoinCount == 0)
+			{
+				throw new InvalidOperationException("Coin selection failed to return a valid coin set.");
+			}
+
+			// How many inputs do we want to provide to the mix?
+			int inputCount = ConsolidationMode ? MaxInputsRegistrableByWallet : GetInputTarget(filteredCoins.Length);
+
+			// Select a group of coins those are close to each other by Anonimity Score.
+			List<IEnumerable<SmartCoin>> groups = new();
+
+			// I can take more coins those are already reached the minimum privacy threshold.
+			for (int i = 0; i < nonPrivateCoinCount; i++)
+			{
+				var group = filteredCoins.Skip(i).Take(inputCount);
+				groups.Add(group);
+			}
+
+			// Calculate the anonScore cost of input consolidation.
+			List<(long Cost, IEnumerable<SmartCoin> Group)> anonScoreCosts = new();
+			foreach (var group in groups)
+			{
+				var smallestAnon = group.Min(x => x.HdPubKey.AnonymitySet);
+				var cost = 0L;
+				foreach (var coin in group.Where(c => c.HdPubKey.AnonymitySet != smallestAnon))
+				{
+					cost += (coin.Amount.Satoshi * coin.HdPubKey.AnonymitySet) - (coin.Amount.Satoshi * smallestAnon);
+				}
+
+				anonScoreCosts.Add((cost, group));
+			}
+
+			var bestCost = anonScoreCosts.Select(g => g.Cost).Min();
+			var bestgroups = anonScoreCosts.Where(x => x.Cost == bestCost).Select(x => x.Group);
+
+			// Select the group where the less coins coming from the same tx.
+			var bestgroup = bestgroups
+				.Select(group =>
+					(Reps: group.GroupBy(x => x.TransactionId).Sum(coinsInTxGroup => coinsInTxGroup.Count() - 1),
+					Group: group))
+				.MinBy(i => i.Reps).Group;
```

Comment:
Can you add ToShuffle here?

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/6993/commits/dcaccec90b2120d1115da32ba4dae374a5f2b188